strapi/strapi
# enhancement: apollo landing page config

### What does it do?

- marks `playgroundAlways` as deprecated, to be removed in Strapi v6
- adds graphql plugin `landingPage` config value that accepts either
  -  a boolean value
  - a user-defined function that returns either
    - a boolean
    - an `ApolloServerPlugin` that implements `renderLandingPage` (see https://www.apollographql.com/docs/apollo-server/api/plugin/landing-pages )

with the meanings:
- `true` means show the "local landing page" (formerly playground)
- `false` means to show the "production landing page" (the generic page with the graphql url but no 'playground')
- An `ApolloServerPlugin` means to use that directly instead of either of the built-in options
- if both `playgroundAlways` and `landingPage` values are left undefined, use `false` for production environments and `true` for all other environments

### Why is it needed?

- Apollo removed 'playground' in Apollo v4 so it's a bad name for it
- We need a way to disable it everywhere, and `playgroundAlways: false` doesn't logically make sense, it would just mean to not always enable it
- This adds the ability for users to provide whatever Apollo landing page plugin they wish, or dynamically determine whether to show it or not, so in theory they can put whatever they want there

QUESTION: Should this be called a new `feat` so it triggers a minor version since it flags playgroundAlways as deprecated? (although we won't change how it actually works until Strapi 6)

### How to test it?

- play around with the `playgroundAlways` values and `landingPage` values and see that it works as expected
- note that `playgroundAlways:true` will always take precedence to avoid a breaking change

### Related issue(s)/PR(s)

fixes https://github.com/strapi/strapi/issues/21018

Docs PR: https://github.com/strapi/documentation/pull/2301
